### PR TITLE
Use correct mask for packed inputs in Qwen-VL 

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As per title, gets rid of `if/else` per attn implementation

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_vl, qwen2_5_vl

**Comment by zucchini-nlp** [comment]
run-slow: ernie4_5_vl_moe, glm4v, glm4v_moe, glm_ocr, paddleocr_vl, qwen2_5_omni, qwen2_5_vl, qwen2_vl, qwen3_5, qwen3_5_moe, qwen3_omni_moe, qwen3_vl, qwen3_vl_moe, video_llama_3

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py [hunk 1]
```diff
@@ -585,7 +585,7 @@ def forward(
         self,
         hidden_states: torch.Tensor,
         cu_seqlens: torch.Tensor,
-        rotary_pos_emb: torch.Tensor | None = None,
+        attention_mask: torch.Tensor | None = None,
```
Comment (zucchini-nlp) [inline src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:588]: unused arg, prob was a bad copy from other models. We use only `position_embeddings`

Comment (vasqu) [reply src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:588]: Oops, very likely yes 😓 I started from qwen vl and opted to make it a bit cleaner back then iirc

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py [hunk 2]
```diff
@@ -603,47 +603,22 @@ def forward(
...
-                    scaling=self.scaling,
-                    dropout=0.0 if not self.training else self.attention_dropout,
-                    is_causal=False,
-                    **kwargs,
-                )[0]
-                for q, k, v in zip(*splits)
-            ]
-            attn_output = torch.cat(attn_outputs, dim=1)
+        max_seqlen = (cu_seqlens[1:] - cu_seqlens[:-1]).max()  # FA-kwargs
+        attn_output, _ = attention_interface(
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:607]: Just noticing it now, but we never collected the attention weights of the vision side 👀

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py [hunk 3]
```diff
@@ -603,47 +603,22 @@ def forward(
...
+            key_states,
+            value_states,
+            attention_mask=attention_mask,
+            scaling=self.scaling,
+            dropout=0.0 if not self.training else self.attention_dropout,
+            cu_seq_lens_q=cu_seqlens,
+            cu_seq_lens_k=cu_seqlens,
+            max_length_q=max_seqlen,
+            max_length_k=max_seqlen,
+            is_causal=False,
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:619]:
```suggestion
```
I know this comes from previous implementations, but we really shouldn't pass this manually ourselves but rely on `self.is_causal`

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py [hunk 4]
```diff
@@ -952,13 +927,27 @@ def forward(
...
+        seq_lengths = cu_seqlens[1:] - cu_seqlens[:-1]
+        packed_sequence = torch.repeat_interleave(
+            torch.arange(len(seq_lengths), device=hidden_states.device), seq_lengths
+        ).unsqueeze(0)
+
+        attention_mask = create_bidirectional_mask(
+            config=self.config,
+            inputs_embeds=hidden_states[None, ...],
+            attention_mask=None,
+            and_mask_function=packed_sequence_mask_function(packed_sequence),
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:939]: `and_mask_function` is quite expensive on runtime. I'd prefer if we could natively integrate it into our mask API instead, not to use vmap (should work OOB)

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py [hunk 5]
```diff
@@ -603,47 +603,22 @@ def forward(
...
-                    attention_mask=None,
-                    scaling=self.scaling,
-                    dropout=0.0 if not self.training else self.attention_dropout,
-                    is_causal=False,
-                    **kwargs,
-                )[0]
-                for q, k, v in zip(*splits)
-            ]
-            attn_output = torch.cat(attn_outputs, dim=1)
+        max_seqlen = (cu_seqlens[1:] - cu_seqlens[:-1]).max()  # FA-kwargs
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:606]: Unsure how / where max_seqlen would be called down the line, but I remember a lot of `.item()` to match FA signature (e.g. for compile); so might be smarter to call it once and before entering each attention module

**Review by vasqu — COMMENTED** [review]
Some comments from my side, definitely the right way to go! We should aim to natively support this in our mask API (without vmap, i.e. the and mask fn) --> otherwise we lose quite a lot of perf

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:607]: in qwen-vl we did collect as a list of 4D tensors per layer. Prob it wasn't standard and was not copied everywhere

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:939]: hmm, what do you mean by "natively"? I could prepare by looping over cu seq lens and un-masking each block and keep it as a small fn in model file. In that case we don't use `create_bidiractional_mask`

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:606]: yeah, might do in the `VisionModel` once. IIRC fa wouldn't prepare max length for us so we have to pass it explciitly

File: src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py (hunk 4 above)
Comment (vasqu) [reply src/transformers/models/ernie4_5_vl_moe/modeling_ernie4_5_vl_moe.py:939]: If you check the new mask API, we force vmap when we pass and/or mask functions https://github.com/huggingface/transformers/blob/efdcbc77ea10f4eb5106b9a4d6df80c541f10a07/src/transformers/masking_utils.py#L1003-L1017

Natively in this case would mean either a new kwarg that we can use to control packed sequences as well (instead of pos ids) or a new function similar to how sliding window is extended https://github.com/huggingface/transformers/blob/efdcbc77ea10f4eb5106b9a4d6df80c541f10a07/src/transformers/masking_utils.py#L1154

The problem with and/or masks is that we can have no idea about their functionality and vmaping is powerful enough to support almost anything, so we are kinda forced to vmap for compatibility/safety

**Comment by zucchini-nlp** [comment]
Discussed internally, we'll add support for torch varlen
